### PR TITLE
Display tileable progress, status and dependency link type on task detail page

### LINKED ISSUE #2342: Tileable Progress can be illustrated on Task Detail DAG
**Is your feature request related to a problem? Please describe.**
On the Task Detail DAG, we can paint the color of each tileable based on its progress. This will help us to oversee the status of the task. 

Right now, all the tileables on the DAG are filled with a yellow color:
Picture1

 but if we paint the color based on the percentage finished, the graph will be more informative. A tileable that has been executed can have a filled color, while a tileable that is 70% executed can have 70% of it filled with yellow and the other 30% remain white.  This gives us a more straightforward understanding of the progress of the task overall. 

**Describe the solution you'd like**
We can create another API endpoint that takes a task id and returns a list/dict that contains all tileables in the DAG and their percentage finished as a float between 0 and 1, where 1 means this tileable has been executed and 0 means this tileable has not been executed, and a float in between means how much of this tileable has been executed.

A sample JSON returned may look like this: 

`
{
    [
        { tileable_id_1: 1 },
        { tileable_id_2: 0.7 },
        ...
    ]
}
`

Then on the frontend, we can change the color for tileables based on the corresponded float value.

### PR DESCRIPTION
**Pull request by RandomY-2**
## What do these changes do?

Now the tileables will be colored based on their status:
- white = pending
- yellow = running
- green = finished
- gray = cancelled
- red = failed

and the tileables will be colored based on the percentage of subtasks they have finished. A sample graph may look like this:

11629380137_ pic_hd

Lastly, the dependency between two tileables will be dashed if the dependency has a linkType of 1.

## Related issue number

Resolves #2342

### CODE REVIEW

File: mars/services/web/ui/src/task_info/TaskTileableGraph.js [hunk 1]
```diff
@@ -34,12 +34,22 @@ export default class TaskTileableGraph extends React.Component {
...
         const { sessionId, taskId } = this.props;
 
+        fetch(`api/session/${sessionId}/task/${taskId
+        }/tileable_detail`)
+            .then(res => res.json())
+            .then((res) => {
+                this.setState({
+                    tileableDetails: res,
+                })
+            });
```
Comment (wjsi) [inline mars/services/web/ui/src/task_info/TaskTileableGraph.js:51]: Progresses of nodes need to be updated periodically.

File: mars/services/web/ui/src/task_info/TaskTileableGraph.js [hunk 2]
```diff
@@ -58,10 +68,88 @@ export default class TaskTileableGraph extends React.Component {
...
+                && prevStates.dependencies !== this.state.dependencies)
+                || prevStates.tileableDetails !== this.state.tileableDetails) {
             d3Select('#svg-canvas').selectAll('*').remove();
 
+            // Set up an SVG group so that we can translate the final graph.
+            const svg = d3Select('#svg-canvas'),
+                inner = svg.append('g');
+
+            svg
+                .append("circle")
```
Comment (wjsi) [inline mars/services/web/ui/src/task_info/TaskTileableGraph.js:85]: Make a utility function to create these shapes? Code a little redundant for me. Comments also needed to show what is actually done here.

File: mars/services/web/ui/src/task_info/TaskTileableGraph.js [hunk 3]
```diff
@@ -71,19 +159,65 @@ export default class TaskTileableGraph extends React.Component {
                 value.rx = value.ry = 5;
                 this.g.setNode(tileable.tileableId, value);
 
+                const tileableDetail = this.state.tileableDetails[tileable.tileableId];
+
+                var nodeProgressGradient = inner.append('linearGradient')
+                    .attr('id', 'progress-' + tileable.tileableId);
+
+                if (tileableDetail.status === 0) {
```
Comment (wjsi) [inline mars/services/web/ui/src/task_info/TaskTileableGraph.js:167]: Create a lookup table to increase readability of this code? For instance,

```javascript
const statusColors = [
  color,  // pending
];

File: mars/services/web/ui/src/task_info/TaskTileableGraph.js [hunk 4]
```diff
@@ -71,19 +159,65 @@ export default class TaskTileableGraph extends React.Component {
...
+                    nodeProgressGradient.append('stop')
+                    .attr('stop-color', '#ffffff')
+                    .attr('offset', '0%');
```
Comment (wjsi) [inline mars/services/web/ui/src/task_info/TaskTileableGraph.js:170]: Just apply `tileableDetail.progress * 100 + '%'` to all states.

**Review by wjsi — CHANGES_REQUESTED** [review]
Besides comments, three UI improvements needed.

1. please make the DAG centralized horizontally
2. saturation of status colors are too high. I recommend `#E74C3C` for failed, `#00CD95` for succeeded and `#BFC9CA` for cancelled. You may choose other color palettes if they look better.
3. add a box with white background to separate legend from DAG.

File: mars/services/web/ui/src/task_info/TaskTileableGraph.js [hunk 5]
```diff
@@ -51,39 +94,170 @@ export default class TaskTileableGraph extends React.Component {
...
+            .attr('x', textX)
+            .attr('y', textY)
+            .text(text)
+            .style('font-size', '15px')
+            .attr('alignment-baseline', 'middle');
+    }
+
     /* eslint no-unused-vars: ["error", { "args": "none" }] */
     componentDidUpdate(prevProps, prevStates, snapshot) {
+        if (Object.keys(this.state.tileableDetails).length !== this.state.tileables.length) {
```
Comment (wjsi) [inline mars/services/web/ui/src/task_info/TaskTileableGraph.js:167]: tileables may never change, but progresses of tileables may change.

Comment (RandomY-2) [reply mars/services/web/ui/src/task_info/TaskTileableGraph.js:165]: https://user-images.githubusercontent.com/69881587/130418781-824da109-5d91-4dd2-9378-7462782396bb.mov

The progress and color of the tileable successfully changed

**Review by wjsi — APPROVED** [review]
LGTM

**Review by qinxuye — APPROVED** [review]
LGTM

**Pull request merged**